tracer/src/Datadog.Trace.ClrProfiler.Native/debugger_method_rewriter.cpp

```diff
@@ -6,13 +6,1005 @@
+        Logger::Info("There are no method probes and lines probes for methodDef", methodHandler->GetMethodDef());
+        return S_OK;
+    }
+
+    Logger::Info("Applying ", methodProbes.size(), " method probes and ", lineProbes.size(),
+                 " line probes on methodDef: ", methodHandler->GetMethodDef());
+    
+    if (!methodProbes.empty() || !lineProbes.empty())
+    {
+        const auto hr = Rewrite(moduleHandler, methodHandler, methodProbes, lineProbes);
+
+        if (FAILED(hr))
+        {
+            // Mark all probes as Error
+            for (const auto& probe : probes)
+            {
+                ProbesMetadataTracker::Instance()->SetProbeStatus(probe->probeId, ProbeStatus::_ERROR);
+            }
+        }
+    }
+
+    return S_OK;
+}
+
+/// <summary>
+/// Performs the following instrumentation on the requested bytecode offset:
+///try
+///{
+///  - Invoke LineDebuggerInvoker.BeginLine with object instance (or null if static method) 
+///  - Calls to LineDebuggerInvoker.LogArg with original method arguments
+///  - Calls to LineDebuggerInvoker.LogLocal with method locals
+///  - LineDebuggerInvoker.EndLine
+///}
+///catch (Exception)
+///{
+///  - Store exception into Exception local
+///}
+/// - Executing the selected sequence point (from the selected bytecode offset)
+/// </summary>
+HRESULT DebuggerMethodRewriter::CallLineProbe(
+    const int instrumentedMethodIndex, 
+    CorProfiler* corProfiler, 
+    ModuleID module_id, 
+    ModuleMetadata& module_metadata, 
+    FunctionInfo* caller, 
+    DebuggerTokens* debuggerTokens, 
+    mdToken function_token, 
+    bool isStatic, 
+    std::vector<TypeSignature>& methodArguments, 
+    int numArgs, 
+    ILRewriter& rewriter, 
+    std::vector<TypeSignature>& methodLocals, 
+    int numLocals, 
+    ILRewriterWrapper& rewriterWrapper, 
+    ULONG lineProbeCallTargetStateIndex, 
+    std::vector<EHClause>& lineProbesEHClauses, 
+    const std::vector<ILInstr*>& branchTargets, 
+    const std::shared_ptr<LineProbeDefinition>& lineProbe)
+{
+    const auto& lineProbeId = lineProbe->probeId;
+    const auto& bytecodeOffset = lineProbe->bytecodeOffset;
+    const auto& probeLineNumber = lineProbe->lineNumber;
+    const auto& probeFilePath = lineProbe->probeFilePath;
+
+    ILInstr* lineProbeFirstInstruction;
+    auto hr = rewriter.GetInstrFromOffset(bytecodeOffset, &lineProbeFirstInstruction);
+
+    if (FAILED(hr))
+    {
+        // Note we are not sabotaging the whole rewriting upon failure to lookup for a specific bytecode offset.
+        // TODO Upon implementing the Probe Statuses, we'll need to reflect that.
+        return S_OK;
+    }
+
+    if (lineProbeFirstInstruction->m_opcode == CEE_NOP || lineProbeFirstInstruction->m_opcode == CEE_BR_S ||
+        lineProbeFirstInstruction->m_opcode == CEE_BR)
+    {
+        lineProbeFirstInstruction = lineProbeFirstInstruction->m_pNext;
+    }
```

Comment:
can you clarify a little bit on this? why only these 2 branching opcodes and not the many other ones?

Reply:
Great explanation, thank you. Yes, it’s harmless, I think you can leave it.